### PR TITLE
Evaluate and enhance cloud/edge performance though QUIC 

### PR DESCRIPTION
**Pull request by kexun**
**What type of PR is this?**
> /kind feature

**What this PR does / why we need it**:
1. increase quic protocol support for edgehub
2. increase quic protocol support for cloudhub. cloudhub can both support websocket and quic client incoming connection.
3. Code optimization, give full play to the performance benefits of the QUIC protocol.

**Which issue(s) this PR fixes**:

Fixes #<Evaluate and enhance cloud/edge performance though QUIC>

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by sids-b** [comment]
@kexun any dependency you add in vendor, please also add an entry in this file manually https://github.com/kubeedge/kubeedge/blob/master/external-dependency.md . Thanks

**Comment by kexun** [comment]
> @kexun , please resolve conflicts , gofmt , golint errors , also e2e tests are failing , please take @pavan187 help.

**Comment by kexun** [comment]
> > @kexun , please resolve conflicts , gofmt , golint errors , also e2e tests are failing , please take @pavan187 help.

File: cloud/edgecontroller/conf/controller.yaml [hunk 1]
```diff
@@ -8,8 +8,12 @@ controller:
     node_update_frequency: 10
     kubeconfig: ""   #Enter path to kubeconfig file to enable https connection to k8s apiserver
 cloudhub:
+  protocol_websocket: true # enable websocket protocol
+  port: 10000 # open port for websocket server
+  protocol_quic: true # enable quic protocol
```
Comment (sids-b) [inline cloud/edgecontroller/conf/controller.yaml]: if both websocket and quic are true, will cloudHub start both the servers ?  So the edge decides to connect to which server ?

File: cloud/edgecontroller/pkg/cloudhub/common/io/hubio.go [hunk 2]
```diff
@@ -33,6 +36,7 @@ func (io *JSONWSIO) SetWriteDeadline(time time.Time) error {
 
 // ReadData read data from connection
 func (io *JSONWSIO) ReadData(d interface{}) (int, error) {
+
```
Comment (sids-b) [inline cloud/edgecontroller/pkg/cloudhub/common/io/hubio.go]: please remove extra blank line

File: edge/pkg/edgehub/clients/quicclient/quicclient.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
+package quicclient
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"github.com/kubeedge/viaduct/pkg/api"
```
Comment (sids-b) [inline edge/pkg/edgehub/clients/quicclient/quicclient.go]: please arrange the imports.

File: edge/pkg/edgehub/config/config.go [hunk 4]
```diff
@@ -61,6 +76,11 @@ func init() {
 		log.LOGGER.Errorf("Error in loading Controller configurations in edge hub:  %v", err)
 		panic(err)
 	}
+	err = getQuicConfig()
```
Comment (sids-b) [inline edge/pkg/edgehub/config/config.go]: the function getQuicConfig doesn't return error. Hardcoded return nil is present. We need to add error returns.

Comment (sids-b) [reply edge/pkg/edgehub/config/config.go]: On a second thought, if user wants to use websocket, he may not give quic config and vice versa. The check needs to be done basd on protocol used which is given in configuration.

File: edge/pkg/edgehub/config/config.go [hunk 5]
```diff
@@ -168,3 +195,51 @@ func getExtendHeader() http.Header {
 
 	return header
 }
+
+func getQuicConfig() error {
```
Comment (sids-b) [inline edge/pkg/edgehub/config/config.go]: please return error wherever applicable in this function.

File: edge/pkg/edgehub/config/config_test.go [hunk 6]
```diff
@@ -50,6 +50,17 @@ type webSocketConfigYaml struct {
 	WriteDeadline    string `yaml:"write-deadline,omitempty"`
 }
 
+//webSocketConfigYaml is a structure which is used to generate the test YAML file to test WebSocket config components
```
Comment (sids-b) [inline edge/pkg/edgehub/config/config_test.go]: nit: webSocketConfigYaml -> quicConfigYaml

File: cloud/edgecontroller/conf/controller.yaml (hunk 1 above)
Comment (kexun) [reply cloud/edgecontroller/conf/controller.yaml]: yes, cloudhub support both quic and websocket. and edge decides to connect to which server.

**Comment by sids-b** [comment]
The implementation looks good to me. But we are introducing new configurations/protocol implementations.
Requesting review from @qizha @CindyXing.

**Comment by sids-b** [comment]
This PR refers to #191

**Comment by m1093782566** [comment]
@kexun 

Prior to review or merge this PR, I think we want to see the design document first. Please reference here: https://github.com/kubeedge/kubeedge/tree/master/docs/proposals

**Comment by sids-b** [comment]
@kexun , please fix CI (cloudlint and edge test(UT))

File: docs/proposals/quic-design.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
+---
```
Comment (sids-b) [inline docs/proposals/quic-design.md]: please raise a seperate PR for design document.

File: docs/proposals/quic-design.md [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+
+# Enhanced Cloud/Edge Communication
+
+## Abstract
+In order to enhance cloud and edge communication efficiency, we introduced the QUIC protocol for cloudhub and edgehub.
+I propose that cloudHub should support both websocket and QUIC protocol access at the same time. And the edgehub can choose one of the protocols to access to the cloudhub.
+QUIC is a new transport which reduces latency compared to that of TCP. On the surface, QUIC is very similar to TCP+TLS+HTTP/2 implemented on UDP. Because TCP is implemented in operating system kernels, and middlebox firmware,making significant changes to TCP is next to impossible.However,since QUIC is built on top of UDP,it suffers from no such limitations.
+Key features of QUIC:
+
+* Dramatically reduced connection establishment time
```
Comment (sids-b) [inline docs/proposals/quic-design.md]: Can you please add some statistical numbers in the design doc ??

**Comment by sids-b** [comment]
@kexun , please rebase to latest fork. I think most of the dependencies are added now by a different PR.

**Comment by sids-b** [comment]
@kexun please rebase the fork

**Comment by rohitsardesai83** [comment]
@kexun , let's fix any issues we get while doing the performance tests . Merging the code for now.

**Pull request merged**